tokio/src/runtime/thread_pool/queue.rs

```diff
@@ -0,0 +1,550 @@
+
+    /// Linked-list tail
+    tail: Option<NonNull<task::Header>>,
+}
+
+unsafe impl<T: task::ScheduleSendOnly> Send for Inner<T> {}
+unsafe impl<T: task::ScheduleSendOnly> Sync for Inner<T> {}
+unsafe impl<T: task::ScheduleSendOnly> Send for Inject<T> {}
+unsafe impl<T: task::ScheduleSendOnly> Sync for Inject<T> {}
+
+#[cfg(not(loom))]
+const LOCAL_QUEUE_CAPACITY: usize = 256;
+
+// Shrink the size of the local queue when using loom. This shouldn't impact
+// logic, but allows loom to test more edge cases in a reasonable a mount of
+// time.
+#[cfg(loom)]
+const LOCAL_QUEUE_CAPACITY: usize = 2;
+
+const MASK: usize = LOCAL_QUEUE_CAPACITY - 1;
+
+/// Create a new local run-queue
+pub(super) fn local<T: 'static>() -> (Steal<T>, Local<T>) {
+    debug_assert!(LOCAL_QUEUE_CAPACITY >= 2 && LOCAL_QUEUE_CAPACITY.is_power_of_two());
+
+    let mut buffer = Vec::with_capacity(LOCAL_QUEUE_CAPACITY);
+
+    for _ in 0..LOCAL_QUEUE_CAPACITY {
+        buffer.push(CausalCell::new(MaybeUninit::uninit()));
+    }
+
+    let inner = Arc::new(Inner {
+        head: AtomicU32::new(0),
+        tail: AtomicU32::new(0),
+        buffer: buffer.into(),
+    });
+
+    let local = Local {
+        inner: inner.clone(),
+        next: None,
+    };
+
+    let remote = Steal(inner);
+
+    (remote, local)
+}
+
+impl<T> Local<T> {
+    /// Returns true if the queue has entries that can be stealed.
+    pub(super) fn is_stealable(&self) -> bool {
+        !self.inner.is_empty()
+    }
+
+    /// Returns true if the queue has an unstealable entry.
+    pub(super) fn has_unstealable(&self) -> bool {
+        self.next.is_some()
+    }
+
+    /// Push a task to the local queue. Returns `true` if a stealer should be
+    /// notified.
+    pub(super) fn push(&mut self, task: task::Notified<T>, inject: &Inject<T>) -> bool {
+        let prev = self.next.take();
+        let ret = prev.is_some();
+
+        if let Some(prev) = prev {
+            self.push_back(prev, inject);
+        }
+
+        self.next = Some(task);
+
+        ret
+    }
+
+    /// Pushes a task to the back of the local queue, skipping the LIFO slot.
+    pub(super) fn push_back(&mut self, mut task: task::Notified<T>, inject: &Inject<T>) {
+        loop {
+            let head = self.inner.head.load(Acquire);
+
+            // safety: this is the **only** thread that updates this cell.
+            let tail = unsafe { self.inner.tail.unsync_load() };
+
+            if tail.wrapping_sub(head) < LOCAL_QUEUE_CAPACITY as u32 {
+                // Map the position to a slot index.
+                let idx = tail as usize & MASK;
+
+                self.inner.buffer[idx].with_mut(|ptr| {
+                    // Write the task to the slot
+                    unsafe {
+                        ptr::write((*ptr).as_mut_ptr(), task);
+                    }
+                });
+
+                // Make the task available
+                self.inner.tail.store(tail.wrapping_add(1), Release);
```

Comment:
Release is needed here because stealers may also read `tail`?

Reply:
Yes, stealers acquire `tail`. I added a note.